src/array.jl

```diff
@@ -331,29 +331,77 @@ end
 size(A::CategoricalArray) = size(A.refs)
 Base.IndexStyle(::Type{<:CategoricalArray}) = IndexLinear()
 
+function update_refs!(A::CategoricalArray, newlevels::AbstractVector)
+    oldlevels = levels(A)
+    levelsmap = similar(A.refs, length(oldlevels)+1)
+    # 0 maps to a missing value
+    levelsmap[1] = 0
+    levelsmap[2:end] .= something.(indexin(oldlevels, newlevels), 0)
+
+    refs = A.refs
+    @inbounds for (i, x) in enumerate(refs)
+        refs[i] = levelsmap[x+1]
+    end
+    A
+end
+
+function merge_pools!(A::CatArrOrSub,
+                      B::Union{CategoricalValue, CatArrOrSub};
+                      updaterefs::Bool=true)
+    if isordered(A) && length(pool(A)) > 0 && pool(B) ⊈ pool(A)
+        lev = A isa CategoricalValue ? get(B) : first(setdiff(levels(B), levels(A)))
+        throw(OrderedLevelsException(lev, levels(A)))
+    end
+    newpool = merge_pools(pool(A), pool(B))
+    oldlevels = levels(A)
+    newlevels = levels(newpool)
+    ordered = isordered(newpool)
+    if isordered(A) != ordered
+        A isa SubArray && throw(ArgumentError("cannot set ordered=$ordered on dest SubArray as it would affect the parent. Found when trying to set levels to $newlevels."))
+        ordered!(A, ordered)
+    end
+    pA = A isa SubArray ? parent(A) : A
+    # If A's levels are an ordered superset of new (merged) pool, no need to recompute refs
+    if updaterefs &&
+        (length(newlevels) < length(oldlevels) ||
+         view(newlevels, 1:length(oldlevels)) != oldlevels)
+        update_refs!(pA, newlevels)
+    end
+    pA.pool = newpool
+    A
+end
+
 @inline function setindex!(A::CategoricalArray, v::Any, I::Real...)
     @boundscheck checkbounds(A, I...)
+    # TODO: use a global table to cache subset relations for all pairs of pools
```

Comment:
Should we have also a `setindex!` for the case of multiple selections like `dst[idx_set] = src`? As it seems currently it would be expensive if `src` is a categorical array (we would have to check the pool at every iteration).

This can be left for later as it is performance related only.

Reply:
Right. We could only cover simple cases (with a single set of indices), and then we could probably just call `copyto!`. I wonder whether that optimization should live in Base, as it already has a special method for `Array` that calls `unsafe_copyto!`. If `unsafe_copyto!` fell back to `copyto!`, all arrays would automatically benefit from any optimized implementation of `copyto!`.

Anyway that's for future work I think.

Reply:
I don't quite follow these two comments, but I'd be slightly wary of using a global table as mentioned in the comment due to multi-threaded scenarios. Not sure if that's the kind of global table you have in mind, but just a thought.

Reply:
Well this idea/need isn't new, it just appears in more places. I'm not sure what would be the best solution for multithreading, but if possible I guess a lock wouldn't be worse than the current code which has to check all levels of both pools on each assignment. Mutating pools during a multithreaded operation should be quite rare in practice, so it would be too bad if it affected performance for common cases. Any ideas? Maybe worth discussing this in a separate issue.